### ISSUE TITLE
sometimes one of the tests is hanging

### ISSUE THREAD

**Issue by simsong** [sleuthkit/sleuthkit#3001]
Thoughts @uckelman-sf ?

https://github.com/sleuthkit/sleuthkit/actions/runs/11150480178/job/30991707724

**Comment by uckelman-sf** [comment]
I don't have any ideas atm. Nothing unusual was logged. Possibly this is a transient problem with the job runner?

**Comment by simsong** [comment]
No idea. Let's see what happens next time.

**Comment by simsong** [comment]
happened again:
https://github.com/sleuthkit/sleuthkit/actions/runs/11172561508/job/31059210985

**Comment by simsong** [comment]
I'm re-running https://github.com/sleuthkit/sleuthkit/actions/runs/11172561508 to see if that solves the problem, and I'm setting the job timeout to 30 min, which should be enough.

**Comment by simsong** [comment]
Here is the output of are recent run in which address-sanitizer failed:
https://github.com/sleuthkit/sleuthkit/actions/runs/11196499434/job/31125415741
![image](https://github.com/user-attachments/assets/bc79238c-8ac5-44fc-ab87-2f18ddb1019e)

And here is the output where it worked:
https://github.com/sleuthkit/sleuthkit/actions/runs/11195090186/job/31122341504
![image](https://github.com/user-attachments/assets/3250ebfe-dc77-4edf-91e9-88ddb522e0c7)

@uckelman-sf  - This doesn't make sense to me. What's happening after the `PASS: tests/fiwalk_test` runs that could sometimes hang?

**Comment by joachimmetz** [comment]
Maybe related https://github.com/joachimmetz/sleuthkit/actions/runs/11201164749/job/31135761436#step:12:2865 could be a network connectivity issue

**Comment by simsong** [comment]
https://github.com/sleuthkit/sleuthkit/issues/3022

**Comment by simsong** [comment]
Still, I wonder what's failing. What happens after it prints the last `PASS` before it prints `Testsuite summary` ?

**Comment by joachimmetz** [comment]
asan can slow things significantly as well, not sure how long a local run takes, also realize that the CI test might run with different conditions https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners

**Comment by simsong** [comment]
> asan can slow things significantly as well, not sure how long a local run takes, also realize that the CI test might run with different conditions https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners

But no more programs are running after the last `PASS` prints!
And this is an intermittent problem.

**Comment by simsong** [comment]
It's not clear what is hanging, but chatGPT says that after all of the tests run, automake does some process cleanup and waits for all processes to finish. So it's possible that one of the tests is forking a process that isn't cleaning up. I thought it might be test_fiwalk, but fiwalk only spawns subprocesses when it is doing metadata extraction with plugins, or when libmagic isn't installed and the system needs to use the `file` command for magic.

**Comment by simsong** [comment]
@uckelman-sf  — See https://github.com/sleuthkit/sleuthkit/actions/runs/11467124445/job/31909322117
We now have a case of test_fiwalk running but the `make test` that includes fiwalk of not ending. 
Of course, this is inconclusive because the hang is inconsistent. 
However, I think that I'm going to remove the `test_fiwalk` from the `make tests` and just run it manually. What do you think?

**Comment by uckelman-sf** [comment]
Determining the cause of the problem is preferable to papering it over. It should be possible to find out what process is hanging.

**Comment by simsong** [comment]
I have tried. If you have any suggestions, please let me know. It’s
complicated by the fact that it is indeterminate. Which makes it seem like
it’s probably something with multiple processes running at the same time.

**Comment by uckelman-sf** [comment]
I'll come back to this after I'm done with #3052.

**Comment by uckelman-sf** [comment]
I've added a diagnostic so we can see what's happening next time: https://github.com/sleuthkit/sleuthkit/pull/3052/commits/07c1a5cf4247cdc5e0f0ed5e7253c949d681772e

**Comment by uckelman-sf** [comment]
https://github.com/sleuthkit/sleuthkit/actions/runs/11497739668/job/32002063142?pr=3052

Apparently the runner is hanging:
```
  PID TTY          TIME CMD
    814 ?        00:00:00 systemd
    827 ?        00:00:00 (sd-pam)
   1015 ?        00:00:00 podman pause
   1443 ?        00:00:00 dbus-daemon
   1573 ?        00:00:02 Runner.Listener
   1590 ?        00:00:10 Runner.Worker
  26298 ?        00:00:00 bash
  26299 ?        00:00:00 bash
  26300 ?        00:00:00 make
  30997 ?        00:00:00 make
  31005 ?        00:00:00 bash
  31014 ?        00:00:00 make
  31034 ?        00:00:00 bash
  31063 ?        00:00:00 test_runner.sh
  31067 ?        00:00:00 runner
  31618 ?        00:00:00 ps
```

**Comment by simsong** [comment]
Hm. And not fiwalk. Well, that's progress, I guess. But harder to debug.

**Comment by simsong** [comment]
Does it run all the test simultaneously in different threads, or does it
run them sequential? I remember that being an option.

**Comment by uckelman-sf** [comment]
The tests run sequentially. Catch2 handles running tests in parallel by sharding the tests, not running them in parallel in one process.

**Comment by uckelman-sf** [comment]
```
    PID TTY          TIME CMD
    801 ?        00:00:00 systemd
    816 ?        00:00:00 (sd-pam)
    987 ?        00:00:00 podman pause
   1543 ?        00:00:00 dbus-daemon
   1680 ?        00:00:02 Runner.Listener
   1696 ?        00:00:08 Runner.Worker
  26374 ?        00:00:00 bash
  26375 ?        00:00:00 bash
  26376 ?        00:00:00 make
  31025 ?        00:00:00 make
  31033 ?        00:00:00 bash
  31042 ?        00:00:00 make
  31096 ?        00:00:00 bash
  31124 ?        00:00:00 test_runner.sh
  31129 ?        00:00:00 runner
  31549 ?        00:00:00 ps
0.003 s: void ErrorsTest::testInitialState()
0.013 s: void ErrorsTest::testLengthChecks()
```
The next test in order is `ErrorsTest::testMultithreaded()` I suspect there's a race condition in it which sometimes causes a deadlock.

**Comment by simsong** [comment]
That would make sense.
I've had inconsistent luck with threadSanitizer, but we might want to give it a try. I had previously looked at fs_thread_test.cpp and had concerns. Here is its comment:

```
// This file implements a thread test for the fs layer.  The program
// opens a file system and then launches N threads.  Each thread walks
// through the same shared file system (TSK_FS_INFO) and produces an
// output file named "thread-N.log".  The actual format of the output
// doesn't really matter.  All we need is for the output to be
// different if we hit race conditions.
//
// To turn this into a thread test, the caller (e.g. Makefile or
// script) should arrange to run the program as follows:
//
//   run with one thread; produce thread-0.log; rename to base.log
//   run with N threads; produce thread-0.log, thread-1.log, etc.
//   diff base.log thread-0.log
//   diff base.log thread-1.log
//   ...
//
// The test passes if all of the thread-N.log files are identical to
// the base.log file.  Of course, this does not guarantee thread
// safety, but by running enough threads and enough repetitions of
// the test without error, you can be more confident.
```

**Comment by uckelman-sf** [comment]
I wonder if this is the problem?
```
REQUIRE (sem_post(&shared.sync_barrier) == 0);
// wait for thread to set some things.
REQUIRE (sem_wait(&shared.sync_barrier) == 0);
```
The main thread is using the same semaphore to signal the child thread to run as the child thread uses to signal the main thread that it's finished. What happens if the order of execution is
1. `sem_post` runs on main thread.
2. `sem_wait` runs on main thread.
3. ` sem_wait` is checked on child thread.

So what happens is we increment the semaphore, decrement the semaphore, and then cause the child to wait forever because the semaphore is now at zero again.

Most of the time step 3 runs immediately after step 1, so when step 2 runs in the child the semaphore is in its intended state and the child can proceed.

I believe the immediate solution is for the the parent and child each to have their own mutex they set to signal the other; that way, the order of execution ceases to matter.

I am generally dubious about the threaded code in Sleuthkit, though, and I'm presently unsure why there should be anything multithreaded about error reporting at all.

**Comment by simsong** [comment]
Thank you for looking into this

**Comment by uckelman-sf** [comment]
Closed by #3062.
